src/rgw/rgw_rados.cc

```diff
@@ -8007,36 +8033,67 @@ int RGWRados::cls_bucket_list_ordered(RGWBucketInfo& bucket_info,
       ++count;
     }
 
-    // Refresh the candidates map
+    // refresh the candidates map
     candidates.erase(candidates.begin());
     ++vcurrents[pos];
     if (vcurrents[pos] != vends[pos]) {
       candidates[vcurrents[pos]->first] = pos;
-    }
-  }
+    } else if (list_results[pos].is_truncated) {
+      if (0 == additional_reads--) {
+	ldout(cct, 10) << "RGWRados::" << __func__ << ": returning " <<
+	  count << "entries early as we used all " <<
+	  allowed_additional_reads << " additional reads" << dendl;
```

Comment:
what happens if we hit the limit on additional reads?

Reply:
This is called by RGWRados::Bucket::List::list_objects_ordered, which will re-call this function if it gets fewer than requested number of entries (and not truncated). That will refresh all shards.

Was chatting with @cbodley and he was chatting with @yehudasa who said that the S3 protocol allows fewer than what was asked for to be returned. So we could legally return whatever we were able to get.

Reply:
interesting--big help